src/core/meshcop/dataset_manager.cpp

```diff
@@ -332,21 +334,49 @@ void DatasetManager::SendSet(void)
     }
 }
 
-void DatasetManager::HandleCoapResponse(void *               aContext,
-                                        otMessage *          aMessage,
-                                        const otMessageInfo *aMessageInfo,
-                                        Error                aError)
+void DatasetManager::HandleMgmtSetResponse(void *               aContext,
+                                           otMessage *          aMessage,
+                                           const otMessageInfo *aMessageInfo,
+                                           Error                aError)
 {
-    OT_UNUSED_VARIABLE(aMessage);
-    OT_UNUSED_VARIABLE(aMessageInfo);
-    OT_UNUSED_VARIABLE(aError);
-
-    static_cast<DatasetManager *>(aContext)->HandleCoapResponse();
+    static_cast<DatasetManager *>(aContext)->HandleMgmtSetResponse(
+        *static_cast<Coap::Message *>(aMessage), *static_cast<const Ip6::MessageInfo *>(aMessageInfo), aError);
 }
 
-void DatasetManager::HandleCoapResponse(void)
+void DatasetManager::HandleMgmtSetResponse(Coap::Message &aMessage, const Ip6::MessageInfo &aMessageInfo, Error aError)
 {
-    mCoapPending = false;
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    Error    error;
+    StateTlv stateTlv;
+
+    SuccessOrExit(error = aError);
+    VerifyOrExit(Tlv::FindTlv(aMessage, stateTlv) == kErrorNone, error = kErrorFailed);
```

Comment:
nit: Is `kErrorParse` more specific?

Reply:
I'm wondering if it makes sense to return such detail errors to users. I think the user may just want to know whether the request succeeded or failed. Even we return all detailed errors, users may not understand what actually happened. Maybe OpenThread should deal with various internal errors and just return the final result to users. Thus, users will have expected error code which will not be affected by OpenThread's internal error type changes(adding, removing, merging errors). Thoughts?

Reply:
I think it makes sense to give the user a more general error. But here the error is also used for debug purposes. I think here we should firstly get a specific error, then in the following `exit:` section, after logging the error, convert it to a general one and call the callback. Thoughts?

Reply:
SGTM

Reply:
I think it's generally better to hide errors that are not meaningful to callers. 
In this case, I lean slightly towards using `kErrorParse` since it's meaningful for `otDatasetSendMgmtActiveSet` API. 
Let's say the developer calls `otDatasetSendMgmtActiveSet` to send an operational dataset to another Thread device using another Thread implementation. `kErrorParse` gives the developer useful hint that the other Thread device is sending an invalid response.

Reply:
@simonlingoogle I think if another device is misbehaving, a lot of errors could occur, e.g. timeout. Making all of them visible to the API caller would be tedious, and making only the "parse" error visible is inconsistent.

Reply:
> a lot of errors could occur, e.g. timeout

Actually I think we should return OT_ERROR_TIMEOUT to callback. Timeout is a meaningful error to a send request?